src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+
+    console.groupCollapsed(`%cblazor%c Loaded ${toDataSizeString(totalResponseBytes)} resources${linkerDisabledWarning}`, 'background: purple; color: white; padding: 1px 3px; border-radius: 3px;', 'font-weight: bold;', 'font-weight: normal;');
+
+    if (cacheLoadsEntries.length) {
+      console.groupCollapsed(`Loaded ${toDataSizeString(cacheResponseBytes)} resources from cache`);
+      console.table(this.cacheLoads);
+      console.groupEnd();
+    }
+
+    if (networkLoadsEntries.length) {
+      console.groupCollapsed(`Loaded ${toDataSizeString(networkResponseBytes)} resources from network`);
+      console.table(this.networkLoads);
+      console.groupEnd();
+    }
+
+    console.groupEnd();
+  }
+
+  async purgeUnusedCacheEntriesAsync() {
+    // We want to keep the cache small because, even though the browser will evict entries if it
+    // gets too big, we don't want to be considered problematic by the end user viewing storage stats
+    const cachedRequests = await this.cache.keys();
+    const deletionPromises = cachedRequests.map(async cachedRequest => {
+      if (!(cachedRequest.url in this.usedCacheKeys)) {
+        await this.cache.delete(cachedRequest);
+      }
+    });
+
+    return Promise.all(deletionPromises);
+  }
+
+  private async tryValidateAndCacheAsync(name: string, cacheKey: string, expectedContentHash: string, networkResponse: Response): Promise<ArrayBuffer> {
+    const responseBuffer = await networkResponse.arrayBuffer();
+
+    // Now is an ideal moment to capture the performance stats for the request, since it
+    // only just completed and is most likely to still be in the buffer. However this is
+    // only done on a 'best effort' basis. Even if we do receive an entry, some of its
+    // properties may be blanked out if it was a CORS request.
+    const performanceEntry = getPerformanceEntry(networkResponse.url);
+    const responseBytes = (performanceEntry && performanceEntry.encodedBodySize) || undefined;
+    this.networkLoads[name] = { responseBytes };
+
+    // crypto.subtle is only enabled on localhost or HTTPS origins
+    // We only write to the cache if we can validate the content hashes
+    if (typeof crypto !== 'undefined' && !!crypto.subtle) {
+      await assertContentHashMatchesAsync(name, responseBuffer, expectedContentHash);
+
+      // Build a custom response object so we can track extra data such as responseBytes
+      // We can't rely on the server sending content-length (ASP.NET Core doesn't by default)
+      await this.cache.put(cacheKey, new Response(responseBuffer, {
+        headers: {
+          'content-type': networkResponse.headers.get('content-type') || '',
+          'content-length': (responseBytes || networkResponse.headers.get('content-length') || '').toString()
+        }
+      }));
```

Comment:
I don't like the fact that we are storing a different response than the one we got from the network. Could we change that to store the original network response directly? cache.put should allow you to cache the original response in a streaming fashion (that is, the promise will return inmediately and not wait for the entire content to be buffered).

Reply:
It would be best to wait until we've validated the hash before we call `cache.put`, otherwise in some error cases we might stream a response into the cache, then the application might stop running before it gets a chance to validate the hash, detect it was a bad response, and remove it from the cache.

> I don't like the fact that we are storing a different response than the one we got from the network

What's the reason why this would be problematic? We also do want to have a way of storing additional metadata along with the data (see `responseBytes` in the code), and I'm unsure there's a way to mutate the original response to include extra data.

Reply:
It's mostly that we can't predict what will happen. We can change this later if we find it causes issues.